Source/JavaScriptCore/runtime/AtomicsObject.cpp

```diff
@@ -466,6 +474,70 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+JSValue atomicsWaitAsyncImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, unsigned accessIndex, ValueType expectedValue, JSValue timeoutValue)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    ValueType* ptr = typedArray->typedVector() + accessIndex;
+    dataLogF("Thread <%p> typedVector addr: %p\n",  &Thread::current(), typedArray->typedVector());
+
+    double timeoutInMilliseconds = timeoutValue.toNumber(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+    Seconds timeout = Seconds::infinity();
+    if (!std::isnan(timeoutInMilliseconds))
+        timeout = std::max(Seconds::fromMilliseconds(timeoutInMilliseconds), 0_s);
+
+    if (!vm.m_typedArrayController->isAtomicsWaitAsyncAllowedOnCurrentThread()) {
+        throwTypeError(globalObject, scope, "Atomics.waitAsync cannot be called from the current thread."_s);
+        return { };
+    }
+    
+    bool didPassValidation = false;
+    ParkingLot::ParkResult result;
+    {
+        ReleaseHeapAccessScope releaseHeapAccessScope(vm.heap);
+        result = ParkingLot::parkConditionally(
+            ptr,
+            [&] () -> bool {
+                didPassValidation = WTF::atomicLoad(ptr) == expectedValue;
+                return didPassValidation;
+            },
+            [] () { },
+            // [&] () {
+            //     // Vector<Strong<JSCell>> dependencies;
+            //     // dependencies.append(Strong<JSCell>(vm, globalObject));
+            //     // auto ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, WTFMove(dependencies));
+            //     // auto dispatch = [ticket] {
+            //     //     ticket->target()->vm().deferredWorkTimer->scheduleWorkSoon(ticket, [](DeferredWorkTimer::Ticket ticket) {
+            //     //         JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+            //     //         JSGlobalObject* globalObject = jsCast<JSGlobalObject*>(ticket->dependencies[0].get());
+            //     //         VM& vm = globalObject->vm();
+            //     //         promise->resolve(globalObject, jsNontrivialString(vm, "time-out"_s));
+            //     //     });
+            //     // };
+            //     // RunLoop::current().dispatchAfter(timeout, WTFMove(dispatch));
+            // },
+            MonotonicTime::now() + timeout,
+            ParkingLot::Type::WaitAsync);
+    }
+
+    JSValue value;
+    JSObject* obj = constructEmptyObject(getVM(globalObject), globalObject->objectStructureForObjectConstructor());
+    obj->putDirect(vm, vm.propertyNames->async, jsBoolean(true));
+    if (didPassValidation) {
+        JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+        typedArray->possiblySharedBuffer()->addWaiterRecord(accessIndex, promise);
+        dataLogF("Thread <%p> ArrayBuffer* addr: %p\n",  &Thread::current(), typedArray->possiblySharedBuffer());
+        value = JSValue(promise);
```

Comment:
`typedArray->possiblySharedBuffer()` in `waitAsync` provides different `ArrayBuffer*` from `notify`.